zmon_aws_agent/main.py

```diff
@@ -0,0 +1,181 @@
+            if deleted:
+                logger.info('ZMON entity deleted successfully')
+            else:
+                logger.error('Failed to delete entity!')
+
+    return to_be_removed_ids
+
+
+def new_or_updated_entity(entity, existing_entities_dict):
+    # check if new entity
+    if entity['id'] not in existing_entities_dict:
+        return True
+
+    existing_entities_dict[entity['id']].pop('last_modified', None)
+    return entity != existing_entities_dict[entity['id']]
+
+
+def add_new_entities(all_current_entities, existing_entities, zmon_client, json=False):
+    existing_entities_dict = {e['id']: e for e in existing_entities}
+    new_entities = [e for e in all_current_entities if new_or_updated_entity(e, existing_entities_dict)]
+
+    if not json:
+        try:
+            logger.info('Found {} new entities to be added in ZMON'.format(len(new_entities)))
+            for entity in new_entities:
+                logger.info(
+                    'Adding new {} entity with ID: {}'.format(entity['type'], entity['id']))
+                resp = zmon_client.add_entity(entity)
+                logger.info('ZMON response ... {}'.format(resp.status_code))
+        except:
+            logger.exception('Failed to add entity!')
+
+    return new_entities
+
+
+def main():
+    argp = argparse.ArgumentParser(description='ZMON AWS Agent')
+    argp.add_argument('-e', '--entity-service', dest='entityservice')
+    argp.add_argument('-r', '--region', dest='region', default=None)
+    argp.add_argument('-j', '--json', dest='json', action='store_true')
+    argp.add_argument('--no-oauth2', dest='disable_oauth2', action='store_true', default=False)
+    args = argp.parse_args()
+
+    if not args.disable_oauth2:
+        tokens.configure()
+        tokens.manage('uid', ['uid'])
+        tokens.start()
+
+    logging.basicConfig(level=logging.INFO)
+
+    # 1. Determine region
+    if not args.region:
+        logger.info('Trying to figure out region..')
+        try:
+            response = requests.get('http://169.254.169.254/latest/meta-data/placement/availability-zone', timeout=2)
+        except:
+            logger.error('Region was not specified as a parameter and can not be fetched from instance meta-data!')
+            raise
+        region = response.text[:-1]
+    else:
+        region = args.region
+
+    logger.info('Using region: {}'.format(region))
+
+    logger.info('Entity service URL: %s', args.entityservice)
+
+    logger.info('Reading DNS data for hosted zones')
+    aws.populate_dns_data()
+
+    # 2. Get running apps
+    apps = aws.get_running_apps(region)
+
+    infrastructure_account = None
+
+    elbs = []
+    scaling_groups = []
+    rds = []
+    elasticaches = []
+    dynamodbs = []
+
+    new_entities = []
+    to_be_removed = []
+
+    if len(apps) > 0:
+        infrastructure_account = apps[0]['infrastructure_account']
+        elbs = aws.get_running_elbs(region, infrastructure_account)
```

Comment:
can we add all the entities directly to a list? and iterate over the method calls?

something like:
```
for collector_method in (a, b, c):
   x.extend(collector_method(...))